source/slang/slang-compiler-tu.cpp

```diff
@@ -59,27 +61,76 @@ namespace Slang
         CodeGenContext::Shared sharedCodeGenContext(&tp, entryPointIndices, &sink, nullptr);
         CodeGenContext codeGenContext(&sharedCodeGenContext);
 
-        ComPtr<IArtifact> outArtifact;
-        SlangResult res = codeGenContext.emitTranslationUnit(outArtifact);
-
-        sink.getBlobIfNeeded(outDiagnostics);
+        // Mark all public functions as exported, ensure there's at least one
+        bool hasAtLeastOneFunction = false;
 
-        if (res != SLANG_OK)
+        for (auto inst : module->getGlobalInsts())
         {
-            return res;
+            if (inst->getOp() == kIROp_Func)
+            {
+                bool hasResourceType = false;
+
+                // DXIL does not permit HLSLStructureBufferType in exported functions
+                auto type = as<IRFuncType>(inst->getFullType());
+                auto argCount = type->getOperandCount();
+                for (UInt aa = 0; aa < argCount; ++aa)
+                {
+                    auto operand = type->getOperand(aa);
+                    if (operand->getOp() == kIROp_HLSLStructuredBufferType)
+                    {
+                        hasResourceType = true;
+                        break;
+                    }
+                }
+
+                if (!hasResourceType)
+                {
+                    if (isSimpleHLSLDataType(inst))
+                    {
+                        // add HLSL export decoration to inst to preserve it in precompilation
+                        hasAtLeastOneFunction = true;
+                        builder.addDecorationIfNotExist(inst, kIROp_HLSLExportDecoration);
+                    }
+                }
+            }
         }
 
-        ISlangBlob* blob;
-        outArtifact->loadBlob(ArtifactKeep::Yes, &blob);
-
-        auto builder = IRBuilder(module);
-        builder.setInsertInto(module);
-
-        switch (targetReq->getTarget())
+        // Avoid emitting precompiled blob if there are no functions to export
+        if (hasAtLeastOneFunction)
         {
-        case CodeGenTarget::DXIL:
-            builder.emitEmbeddedDXIL(blob);
-            break;
+            ComPtr<IArtifact> outArtifact;
+            SlangResult res = codeGenContext.emitTranslationUnit(outArtifact);
+            sink.getBlobIfNeeded(outDiagnostics);
+            if (res != SLANG_OK)
+            {
+                return res;
+            }
+
+            // Mark all exported functions as available in dxil
+            for (auto inst : module->getGlobalInsts())
```

Comment:
Again, because the final decision will be made during the compilation process, we need a way to associate the linked and legalized IR back to the original module IR and mark the original IR based on the functions in the linked IR.

Reply:
Do you have some idea on how to do that mapping from linkedIR to original module IR?  I'm picturing the IR undergoing lots of cloning and modification during the link and legalization passes.  Would a decoration that held the pointer to the original IRInst survive until the IR was fully linked and optimized?

Reply:
Since I'm already defining a decoration for "attempt to export" named "TransientExportDecoration", I'm now trying to add a pointer operand to it with the payload being the module IR's Inst to remember the mapping from linked IR back to module IR.

Reply:
I tried using a PtrLit operand but it looks like cloneValue() deliberately refuses to copy non-null pointer values.  I'll try making up my own integer->IRInst* mapping

Reply:
I've added an integer operand to the new "TransientExportDecoration" that would allow the mapping of linked IR functions back to the pre-linked and legalized module IR, however, I'm unsure how to correctly do the post-link/legalization filtering and correlation to module IR.

In linkandOptimizeIR(), at the end of the function is where I'm looking to inspect the post-linked IR. I can iterate through the global functions and remove the ones that define or use generics.  Given that, I think I have two approaches.  Either:

1) In linkandOptmizeIR(), after I filter the IR for generics, dig through the module IR and tag things as availableInDXIL that correpond to functions that still exist in the linked IR
or
2) Do the IR filitering at the end of linkAndOptimizeIR() but don't do anything with module IR yet.  Rather, let the function exit and then in precompileForTargets(), dig back into the output artifact and somehow get the linked IR to check for the "TransientExportDecoration" markers.  I don't know the commands to do that if it's currently possible. The output artifact blob contains just DXIL I think.  That is, in precompileForTarget(), I call codeGenContext.emitPrecompiledDXIL(outArtifact), followed by "outartifact->loadBlob()".  Not sure how I would also be able to get a hold of the linked IR.

Reply:
There is already a `IRExportDeocration` that encodes the mangled name of symbols as long as they are considered "exported" in the slang IR sense. Note that this is not the same as `IRHLSLExportDeocration`, which are for symbols marked with `export` in slang code.

All public symbols should automatically have `IRExportDecoration` already.

Reply:
I think you're saying that I don't need to add an integer for the cross referencing and I could use mangled names from IRExportDecoration instead. What about the second part though, which is essentially whether it makes more sense to apply the "availalbleInDXIL" decoration to module IR at the end of linkAndOptimizeIR() or if I can inspect the linked IR results from precompileForTargets() after emitTranslationUnit() is finished producing an output DXIL artifact.

Reply:
I believe either option can work. We can do it in `precomileForTargets()` now for cleanliness, unless we hit any issues with that.

Reply:
Okay, I have something working using the mangled names for mapping, however, I failed to figure out how to correctly extract the linked IR from the output artifact from codeGenContext.emit*() call.

My upcoming patch revision does work, but it uses what I assume is a hack.  I've added a new pointer to IRModule inside TargetProgram.  I'll be looking for advice on how to do that properly.

Reply:
Current revision does the correlation using mangled name in precompileForTarget, however the retrieval of the linked IR is a hack, I believe.

Reply:
Fixed. Getting at linked IR is no longer necessary as the list of exported functions is saved to metadata in the artifact.